### DISCUSSION TITLE
Should we include more dependencies to `TestModule.<Framework>Module`

### DISCUSSION THREAD

**Discussion by lefou** [com-lihaoyi/mill#3281] [Development]
We currently don't add the dependencies to a concrete test runner in our `TestModule` traits.

That means, users need to remember to add the correct dependencies themselves.

The reasons are twofold. 

1. We don't want to maintain version in these traits
2. In some cases it's not easy to determine the best set of dependencies (e.g. scalatest)

This lightweight approach is a bit inconvenient, since you need to look up the correct dependencies every time.

What about adding the dependencies, such that just adding a `extends TestModule.Utest` is enough to run a test? To keep version-update induced maintenance low, we could:

1. Introduce a dedicated `version` target. Either a specific (`TestModule.Utest.utestVersion`, `TestModule.ZioTest.zioTestVersion`) or some generic (`TestModule.Utest.testFrameworkVersion`).

2. Introduce additional targets, to select a specific set of features and deduce the respective needed dependencies from it. (`TestModule.ScalaTest.suiteStyles: T[Seq[SuiteStyle]]`. E.g. to select `free` we add `scalatest-freespec` dependency)

It would be also an opportunity to abstract over some framework specific settings like parallel running of suites/test or failure reporting.

**Comment by lihaoyi** [comment]
For the Java `TestModule`s, we currently do include dependencies. e.g. for Junit4, Junit5, and TestNg:

https://github.com/com-lihaoyi/mill/blob/7147d76254ab18021be81637f515e342b2583d6b/scalalib/src/mill/scalalib/TestModule.scala#L243-L272

**Reply by lefou** [reply]
Yeah, these include the `sbt.testing.framework` (which is uncommon in the Java world), and (transitively) the test runner.
